### DISCUSSION TITLE
Next.js client and php backend

### DISCUSSION THREAD

**Discussion by andrewtheone** [vercel/next.js#13446] [Help]
Hi!

I'm a traditional web developer, i know nodejs and react for modern frontends and I do my backends in mostly php. 

I'm starting a new project and I'd like Next.JS as my frontned/clientside.

While traditionally all session handling is done by the http stack (nginx,apache,php), in hybrid apps its a bit different.

My plan is to have the backend, a restful api server written in php and served under a subdomain (example.com/api/...), and the next.js would be accessible from the root (example.com). 

The application is session heavy, meaning: different responses with different sessions, auth required areas, etc.

My current structure is:
I have my php api app under the /api subfolder, it works great, uppon request it looks for "access_token" cookie or get variable and initializes the session based on the value, if there's no cookie or get variable found, it just creates a new session and returns a set cookie header.

On node side: I have an express http server before app.rendering the nextjs app, it looks for a cookie from the request by the name: 'access_token' if its found, than it validates (creates an api call to: /api/access_token/check) it. (if its not valid, a new session is generated and returned). At this time the expressjs sets the response header's accordingly. It works. In app.render i pass the current access_token value in the 'req' parameter.
In next.js:
_app.js can read the passed in access_token variable in getInitialProps from the req property. It this is the part that's a little bit cloudy to me, currently i'm doing the following:
-I have az API function in ./utils/Api.js with is just a wrapper function:
```js

function Api() {

	const accessToken = Api.store.getState().api.accessToken;
	const dispatch = Api.store.dispatch;
 	const instance = axios.create({
		baseURL: "//example.com/api/"
	});

	return {
		get: function(url, options) {
			if(typeof options === "undefined") {
				options = {};
			}

			if(typeof options.params === "undefined") {
				options.params = {};
			}

			options.params.access_token = accessToken;
			return instance.get(url, options);
		},
		post: function(url, data, options) {
			if(typeof options === "undefined") {
				options = {};
			}

			if(typeof options.params === "undefined") {
				options.params = {};
			}

			options.params.access_token = accessToken;
			return instance.post(url, data, options);
		},
		fetchUser: async function() {
			const resp = await Api().post('/user/fetch_user');

			if(resp.data.result && resp.data.result.user && resp.data.result.company) {
				dispatch({type: 'SET_SESSION', payload: resp.data.result});
			} else {
				dispatch({type: 'SET_SESSION_LOGOUT', payload: null});
			}
		},
		fetchCart: async function() {
			const resp = await Api().get('/cart/get_items');
			dispatch({type: 'SET_CART_ITEMS', payload: resp.data.result});
			return true;
		},
		addItemToCart: async function(offerId, quantity) {
			const resp = await Api().post('/cart/add_item', {offer_id: offerId, quantity: quantity});
			return true;
		}
	}
	return instance;
}
Api.store = undefined;

export default Api;
```
-In _app's getInitialState I store the passed in access_token in redux and set's the API.store to the redux store (so i can getState() and lookup the stored access_token).
-But i have to reassign the Api.store to the redux store in the "render" method of _app, because the getInitialState is executed on the serverside while the render is executed on the client side.

Currently it's working fine, but I feel like there's a better way to achieve this.

Thanks.

**Comment by thanveershah** [comment]


**Reply by tawfek** [reply]
if I understood your question, you can solve this problem by adding proxy in `package.json` :)
```
"proxy": "http://localhost:3000"
```

or 

```
"proxy": {
    "/api/users": {
        "target": "http://localhost:3000"
    }
  }
```
